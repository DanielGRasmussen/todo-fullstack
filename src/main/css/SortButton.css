#sort-direction {
	width: 30px;
	height: 30px;
	border: 1px solid #ccc;
	background-color: #f2f2f2;
	border-radius: 4px;
	position: relative;
	cursor: pointer;
	margin-top: 14px;
	padding: 5px;
}

@media screen and (min-width: 500px) {
	#sort-direction {
		margin: 14px 0 0 10px;
	}
}

#sort-direction img {
	position: absolute;
	width: 26px;
	height: 26px;
	top: 2px;
	left: 2px;
	padding: 5px;
}

#sort-direction:hover {
	outline-color: #999;
	background-color: #e8e8e8;
}

/* Slowly rotate upside down */
@keyframes upside-down {
	from {
		transform: rotate(0);
	}
	to {
		transform: rotate(180deg);
	}
}

/* Slowly rotate downside-up */
@keyframes downside-up {
	from {
		transform: rotate(180deg);
	}
	to {
		transform: rotate(360deg);
	}
}

/* Title messages (title css is in SearchMenu.css) */
#sort-direction::before {
	top: -38px;
	width: 120px;
}

#sort-direction.ascending::before {
	content: "Ascending Sort";
}

#sort-direction.descending::before {
	content: "Descending Sort";
}
