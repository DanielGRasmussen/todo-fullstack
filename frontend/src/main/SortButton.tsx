import "./css/SortButton.css";
import React, { useState } from "react";

interface ISortButtonProps {
	sortOrder: boolean;
	setSortOrder: React.Dispatch<React.SetStateAction<boolean>>;
}

function SortButton({
	sortOrder,
	setSortOrder
}: ISortButtonProps): JSX.Element {
	/* This renders a button to sort a list in either ascending or descending order.
	 *
	 * This component takes two props:
	 * sortOrder: a boolean flag that is ascending/true by default.
	 * setSortOrder: a function used to update the current sort order.
	 *
	 * This component returns a div element with an onClick handler that triggers the handleOnClick function and toggles
	 * the sort order. The div contains an img element with the up arrow icon, which changes based on the current sort
	 * order.
	 */
	const [first, isFirst] = useState(true);

	// up arrow   = ascending  = a to z
	// down arrow = descending = z to a
	const handleOnClick = (): void => {
		const newSortOrder = !sortOrder;
		isFirst(false);
		setSortOrder(newSortOrder);
	};

	const arrowStyle: { animation: string } = {
		animation: sortOrder
			? "downside-up 0.3s linear forwards"
			: "upside-down 0.3s linear forwards"
	};
	// Since it requires the same type.
	const blankStyle: { animation: string } = { animation: "" };

	return (
		<div
			onClick={handleOnClick}
			id="sort-direction"
			className={(sortOrder ? "ascending" : "descending") + " tooltip"}
		>
			{sortOrder ? (
				<img
					src={process.env.PUBLIC_URL + "/assets/up_arrow.svg"}
					alt="Ascending order"
					style={first ? blankStyle : arrowStyle}
				/>
			) : (
				<img
					src={process.env.PUBLIC_URL + "/assets/up_arrow.svg"}
					alt="Descending order"
					style={arrowStyle}
				/>
			)}
		</div>
	);
}

export default SortButton;
