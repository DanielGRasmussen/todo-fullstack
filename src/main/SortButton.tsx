import "./css/SortButton.css";
import React, { useState } from "react";

function SortButton({ onClick }): JSX.Element {
	const [isAscending, setIsAscending] = useState(true);
	const [first, isFirst] = useState(true);

	// up arrow   = ascending  = a to z
	// down arrow = descending = z to a
	const handleOnClick = (): void => {
		setIsAscending(!isAscending);
		isFirst(false);
		onClick(isAscending);
	};

	const arrowStyle: { animation: string } = {
		animation: isAscending
			? "downside-up 0.3s linear forwards"
			: "upside-down 0.3s linear forwards"
	};
	// Since it requires the same type.
	const blankStyle: { animation: string } = { animation: "" };

	return (
		<div
			onClick={handleOnClick}
			id="sort-direction"
			className={(isAscending ? "ascending" : "descending") + " tooltip"}
		>
			{isAscending ? (
				<img
					src={
						process.env.PUBLIC_URL + first
							? "/assets/up_arrow.svg"
							: "/assets/down_arrow.svg"
					}
					alt="Ascending order"
					style={first ? blankStyle : arrowStyle}
				/>
			) : (
				<img
					src={process.env.PUBLIC_URL + "/assets/up_arrow.svg"}
					alt="Descending order"
					style={arrowStyle}
				/>
			)}
		</div>
	);
}

export default SortButton;
