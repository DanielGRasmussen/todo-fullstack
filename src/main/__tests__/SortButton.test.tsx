import SortButton from "../SortButton";
import React from "react";
import { fireEvent, render } from "@testing-library/react";

describe("SortButton", () => {
	it("should render correctly", () => {
		const { container } = render(
			<SortButton sortOrder={true} setSortOrder={jest.fn()} />
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should change sort order when clicked", () => {
		const setSortOrderMock = jest.fn();
		const { getByTestId } = render(
			<SortButton sortOrder={true} setSortOrder={setSortOrderMock} />
		);
		const button = getByTestId("sort-direction");
		fireEvent.click(button);
		expect(setSortOrderMock).toHaveBeenCalledWith(false);
	});
});
