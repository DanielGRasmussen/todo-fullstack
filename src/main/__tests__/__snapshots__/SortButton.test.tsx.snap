// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortButton should render correctly 1`] = `
<div
  class="ascending tooltip"
  id="sort-direction"
>
  <img
    alt="Ascending order"
    src="/assets/up_arrow.svg"
  />
</div>
`;
